### PR TITLE
ShellPkg: Code refactoring

### PR DESCRIPTION
**Pull request by pierregondois**
# Description

In an effort to simplify the code in the ShellPkg:
- Refactor the code of some shell commands
- Remove deprecated/unsused code
- Introduce aliases for the `ShellPrintEx()` and `ShellPrintHiiEx()` functions to remove optional parameters.

For most patches, no functional change is introduced.
 
- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

Compared the output of the following commands with/without the patches:
```
memmap
memmap -sfo

dmem
dmem -verbose

drvcfg
# Compared the content of XXX
drvcfg -o XXX

connect
# output:
# Connect - Handle [17] Result Success.
# output: Connect - Handle [38] Result Success.
# Then
connect 17
connect 38
```

### CODE REVIEW

**Comment by leiflindholm** [comment]
Before I start nitpicking, I just really want to say: thank you for this brilliant bit of cleanup.

File: ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c [hunk 1]
```diff
@@ -504,20 +501,12 @@ DisplaySystemTable (
...
-    if (ShellStatus == SHELL_SUCCESS) {
-      ShellStatus = DisplayConformanceProfiles (ConformanceProfileTableAddress);
-    }
+    DisplayRtProperties (RtPropertiesTableAddress);
+    DisplayImageExecutionEntries (ImageExecutionTableAddress);
+    DisplayConformanceProfiles (ConformanceProfileTableAddress);
   }
 
-  return ShellStatus;
+  return SHELL_SUCCESS;
```
Comment (leiflindholm) [inline ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c]: I agree with this change, but the commit message is actually incorrect. ShellStatus _is_ used, as the return value from each previous function call.

But a) I don't necessarily see great value in that.
And b)
- DisplayRtProperties looks broken for this anyway:
It takes an address in, ignores it, looks up the config table anyway, doesn't check the return value, uses the result blindly, and then only does an error check on whether the printout succeeded...
- DisplayImageExecutionEntries and DisplayConformanceProfiles also ignore the address parameter completely, and forego any useful error handling.

So, please keep this change, but also update the commit message indicating that the change _is_ functional. Just not practically so.

I see you rework these helper functions in later cmomits, so I'll comment on their pre-existing brokenness there, I think there is more cleanup required.

File: ShellPkg/Library/UefiShellDebug1CommandsLib/MemMap.c [hunk 2]
```diff
@@ -147,6 +147,203 @@ AddMemoryLength (
...
+  MEMORY_LENGTH_ENTRY     *Entry;
+
+  InitializeListHead (&MemoryList);
+
+  TotalPages = 0;
+
+  // Reset the MemoryPageCount array.
+  ZeroMem (MemoryPageCount, sizeof (MemoryPageCount));
+
+  for ( Walker = Descriptors
```
Comment (leiflindholm) [inline ShellPkg/Library/UefiShellDebug1CommandsLib/MemMap.c]: Stray leading space after (? Ah, retained from original. But I think it's worth cleaning up.

File: ShellPkg/Library/UefiShellDebug1CommandsLib/MemMap.c [hunk 3]
```diff
@@ -147,6 +147,203 @@ AddMemoryLength (
...
 }
 
+/** Parse Memory Descriptors
+
+@param[in]  Descriptors   Array of pointers to Memory Descriptors.
+@param[in]  Size          Size of the Descriptors array.
+@param[in]  ItemSize      Size of a Memory Descriptor.
+@param[in]  Sfo           Whether 'Standard Format Output' should be used.
+**/
+SHELL_STATUS
```
Comment (leiflindholm) [inline ShellPkg/Library/UefiShellDebug1CommandsLib/MemMap.c:158]: static?

File: ShellPkg/Include/Library/ShellLib.h [hunk 4]
```diff
@@ -968,6 +968,16 @@ ShellPrintHiiEx (
   ...
   );
 
+/** Macro aliasing the ShellPrintEx() function and providing default parameters for
+    the Col and Row position.
+**/
+#define ShellPrintDfltEx(...)  ShellPrintEx(-1, -1, __VA_ARGS__)
```
Comment (leiflindholm) [inline ShellPkg/Include/Library/ShellLib.h]: I'm all for the idea, but Dflt makes for annoying CamelCase. Please expand to Default (for both).

**Review by leiflindholm — CHANGES_REQUESTED** [review]

File: ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c (hunk 1 above)
Comment (leiflindholm) [reply ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c]: Ah, no, that was just the printouts. The actual logic change was preceding...

File: ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c [hunk 5]
```diff
@@ -97,47 +97,44 @@ DisplayRtProperties (
 {
   EFI_RT_PROPERTIES_TABLE  *RtPropertiesTable;
   UINT32                   RtServices;
-  SHELL_STATUS             ShellStatus;
   EFI_STATUS               Status;
 
-  ShellStatus = SHELL_SUCCESS;
-
-  if (Address != 0) {
```
Comment (leiflindholm) [inline ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c:105]: Yeah, so here - if you look at the original code, it never uses Address at all. And it never checks the return value ot EfiGetSystemConfigurationTable.

The pattern repeats across all three functions.
I'd suggest turning this into a functional change:
- dropping the EfiGetSystemConfigurationTable calls (which will already have been performed before calling this function with its address)
- actually using the address
- and then considering whether these need return values at all for a printout

Comment (pierregondois) [reply ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c:105]: Hoping I understood the comment correctly, the changes should have been made

File: ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c [hunk 6]
```diff
@@ -88,27 +88,24 @@ DisplayMmioMemory (
...
-  SHELL_STATUS             ShellStatus;
-  EFI_STATUS               Status;
+  UINT32        RtServices;
+  SHELL_STATUS  ShellStatus;
+  EFI_STATUS    Status;
 
   ShellStatus = SHELL_SUCCESS;
 
-  if (Address == 0) {
+  if (RtPropertiesTable == 0) {
```
Comment (leiflindholm) [inline ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c]: I approve of the changing the "Address" parameters to pointers, but then it also follows that the tests should be for NULL, not 0.

Comment (pierregondois) [reply ShellPkg/Library/UefiShellDebug1CommandsLib/Dmem.c]: yes right indeed

File: ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c [hunk 7]
```diff
@@ -154,9 +154,9 @@
     //
     Status = gBS->StartImage (LoadedDriverHandle, NULL, NULL);
     if (EFI_ERROR (Status)) {
-      ShellPrintHiiEx (-1, -1, NULL, STRING_TOKEN (STR_LOAD_ERROR), gShellLevel2HiiHandle, FileName, Status);
+      ShellPrintHiiDefaultEx (STRING_TOKEN (STR_LOAD_ERROR), gShellLevel2HiiHandle, FileName, Status);
     } else {
-      ShellPrintHiiEx (-1, -1, NULL, STRING_TOKEN (STR_LOAD_LOADED), gShellLevel2HiiHandle, FileName, LoadedDriverImage->ImageBase, Status);
+      ShellPrintHiiDefaultEx (STRING_TOKEN (STR_LOAD_LOADED), gShellLevel2HiiHandle, FileName, LoadedDriverImage->ImageBase, Status);
```
[the remark this answers is not in the document]
Comment (leiflindholm) [reply ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c]: Another pre-existing issue caused by appalling code design (looking like the traditional misintepretation of MISRA). Ultimately, success-handling.

Untangle the success-handling and the error report (for the case that cannot be reached) goes away.

[the remark this answers is not in the document]
Comment (pierregondois) [reply ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c]: Yes ok, I thought it was a false positive

[the remark this answers is not in the document]
Comment (leiflindholm) [reply ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c]: Tbh, I was thinking to take it further:
https://github.com/leiflindholm/edk2/commit/b443101616de1ac30ab8f7257369f69d771d87be

[the remark this answers is not in the document]
Comment (pierregondois) [reply ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c]: yes ok, I replaced the last patch by yours if it's ok

[the remark this answers is not in the document]
Comment (leiflindholm) [reply ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c]: My bad (uncrustify), update pushed to https://github.com/leiflindholm/edk2/tree/rework/11527

**Comment by pierregondois** [comment]
Just a ping in case the PR looks sensible

**Comment by leiflindholm** [comment]
> Just a ping in case the PR looks sensible

Wanted to give others a chance to comment. Thanks for the ping, I've set the push label.
